### PR TITLE
Avoid potential segfault in CSR and CSC matrix indexing

### LINKED ISSUE #11125: BUG: segfault when slicing a CSR or CSC sparse matrix with slice start index > stop index
I noticed a segfault in SciPy that occurs when running the CuPy test suite with 1.4.0rc1. This seems to occur when using a slice with a start index greater than the number of rows to index the matrix.

#### Reproducing code example:

```Python
import numpy as np
import scipy.sparse

data = np.array([0, 1, 2, 3], np.float32)
indices = np.array([0, 1, 3, 2], 'i')
indptr = np.array([0, 2, 3, 4], 'i')
m = scipy.sparse.csr_matrix((data, indices, indptr), shape=(3, 4))

# invalid index that exceeds size of the first axis causes a segfault!
z = m[3:2]

# z = m[3]  # this case raises an IndexError as expected

```

#### Error message:

```
Fatal Python error: Segmentation fault

Current thread 0x00007f28492c5700 (most recent call first):
  File "/home/lee8rx/miniconda3_new/envs/pyir/lib/python3.7/site-packages/scipy/sparse/compressed.py", line 789 in _get_submatrix
  File "/home/lee8rx/miniconda3_new/envs/pyir/lib/python3.7/site-packages/scipy/sparse/compressed.py", line 652 in _get_sliceXslice
  File "/home/lee8rx/miniconda3_new/envs/pyir/lib/python3.7/site-packages/scipy/sparse/_index.py", line 51 in __getitem__
```

#### Scipy/Numpy/Python version information:

```
1.4.0rc1 1.17.2 sys.version_info(major=3, minor=7, micro=3, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by grlee77**
#### Reference issue

closes #11125

#### What does this implement/fix?

This PR guards against a potential segfault that seems to have been around since SciPy 1.3.0.  An `IndexError` will now be raised instead (as occurs for older Scipy <1.3).

### CODE REVIEW

**Comment by perimosocordiae** [comment]
I'll be taking a closer look at this soon, but first a question: do the other sparse matrix classes handle this case correctly? Ideally we'd have a single test case in the big `test_base.py` file that covers the behavior of all formats.

**Comment by grlee77** [comment]
I'll be taking a closer look at this soon, but first a question: do the other sparse matrix classes handle this case correctly? 

I can take a look. Only CSR and CSC were causing the segfaults in the CuPy test suite, but I am not sure if there are equivalent test cases for the other classes.

**Comment by grlee77** [comment]
For other classes:

DIA and COO do not allow slicing at all so they return e.g. (`TypeError: 'coo_matrix' object is not subscriptable`) which seems reasonable.

DOK and LIL do not segault, but return a sparse matrix with shape 0 along the sliced axis.:
e.g.
```Python

In [49]: m_dok = m.todok()

In [50]: m_dok[3:2]

Out[50]: 
<0x4 sparse matrix of type '<class 'numpy.float32'>'
	with 0 stored elements in Dictionary Of Keys format>
```

**Comment by grlee77** [comment]
CSR and CSC also give shape 0 along an axis for slices that are out of range. That might also be considered a bug, but not the one I intended to address in this PR:

```Python
m_csr[5:6]
```
when m_csr is a CSR matrix with shape 3 x 4 gives:
```Python
<0x4 sparse matrix of type '<class 'numpy.float32'>'
	with 0 stored elements in Compressed Sparse Row format>
```

vs.
```Python
m_csr[5]
```
which gives
```Python
Traceback (most recent call last):

  File "<ipython-input-51-ce819d42d4fd>", line 1, in <module>
    m[5]

  File "/home/lee8rx/miniconda3_new/envs/pyir/lib/python3.7/site-packages/scipy/sparse/_index.py", line 35, in __getitem__
    row, col = self._validate_indices(key)

  File "/home/lee8rx/miniconda3_new/envs/pyir/lib/python3.7/site-packages/scipy/sparse/_index.py", line 135, in _validate_indices
    raise IndexError('row index (%d) out of range' % row)

IndexError: row index (5) out of range
```

**Comment by grlee77** [comment]
I guess in order to proceed we need to decide what is the desired behavior for all of these cases. (potentially considering the behavior for older versions of SciPy pre 1.3 as well).
It seems most intuitive to me that these out of range slices should all return `IndexError` rather than size 0 sparse arrays, but perhaps there is some other reason for the current behavior?

**Comment by perimosocordiae** [comment]
Older versions of scipy do return size-0 sparse matrices on out-of-bounds slices, which is consistent with both ndarray and numpy.matrix semantics.

This is also the numpy.matrix behavior for empty slices, including slices where start > stop with a positive step, like the ones that cause the segfault here.

So in this PR I think we should:
 1. add a simple test for all indexable formats that ensures consistency with numpy.matrix
 2. fix the segfault in CSR/CSC formats

**Comment by perimosocordiae** [comment]
In fact, we don't even need a new test. You can add two additional cases here:
https://github.com/scipy/scipy/blob/9bd72b0ef13d5bc3cf700258bac9245cd0769a9b/scipy/sparse/tests/test_base.py#L2386

to test out-of-bounds (`s_[15:20]`) and invalid (`s_[3:2]`) cases.

**Comment by tylerjereddy** [comment]
I revised this PR based on CJ's feedback. Feel free to adjust my revisions as you see fit---just trying to get this moving before `rc2`, which is slated for Sunday if possible.

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge

**Comment by grlee77** [comment]
Thanks @tylerjereddy! I had been travelling and hadn't gotten a chance to revisit this.

File: scipy/sparse/compressed.py [hunk 1]
```diff
@@ -1276,8 +1281,7 @@ def _process_slice(sl, num):
         if stride != 1:
             raise ValueError('slicing with step != 1 not supported')
         if i0 > i1:
-            raise IndexError(
-                'slicing with start index > stop index is not supported')
+            return 'empty', 'empty'
```
Comment (grlee77) [inline scipy/sparse/compressed.py]: This seems like the desired behavior, but is not consistent with older SciPy which did raise an `IndexError` for CSR/CSC in this case. 
e.g. `IndexError: index out of bounds: 0 <= 3 <= 3, 0 <= 2 <= 3, 3 <= 2`

No objections here, but perhaps worth mentioning in the release notes?

Comment (tylerjereddy) [reply scipy/sparse/compressed.py]: I think CJ wanted that behavior as you saw above---I'll make a note to adjust the release notes & then you & CJ can check that I've described it properly when I do the backporting.

**Comment by tylerjereddy** [comment]
Alright, there's a +1 to merge from a core dev & this is targeted at 1.4.0 rc2, which I'd like to get out soon, so merging.

**Comment by tylerjereddy** [comment]
CI failure is unrelated from #11100

**Comment by mdhaber** [comment]
Thanks @grlee77 for this and all the other PRs over the years! BTW, if you'd like your name added to [THANKS.txt](https://github.com/scipy/scipy/blob/master/THANKS.txt), please make an entry [here](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.5.0*).

**Comment by grlee77** [comment]
I know this was already merged, but wanted to point out one other subtle difference in behavior for the solution merged here vs. NumPy array or sparse LIL/DOK behavior.

Slicing CSR with something like `a[3:2, :]` after this PR give an empty CSR matrix of shape = `(0, 0)`, but doing the same for a NumPy array or some other sparse matrix types would give an empty matrix, but with shape `(0, a.shape[1])`. Is this distinction important to maintain?

e.g.
```Python
import numpy as np
a = np.ones((5, 5)); 
print("shape = {}".format(a[3:2].shape))
```

```shape = (0, 5)```

**Comment by WarrenWeckesser** [comment]
@grlee77, thanks for pointing that out.  I would call that inconsistency a bug.

In the following, `a[2:2, :]` and `a[2:1, :]` should both have shape `(0, 4)`.
```
In [19]: a = csr_matrix([[1, 0, 0, 0], [0, 0, 0, 0], [0, 2, 3, 0]])                                                                                                                                  

In [20]: a.A                                                                                                                                                                                         
Out[20]: 
array([[1, 0, 0, 0],
       [0, 0, 0, 0],
       [0, 2, 3, 0]], dtype=int64)

In [21]: a[2:2, :].A                                                                                                                                                                                 
Out[21]: array([], shape=(0, 4), dtype=int64)

In [22]: a[2:1, :].A                                                                                                                                                                                 
Out[22]: array([], shape=(0, 0), dtype=int64)
```

**Comment by tylerjereddy** [comment]
Probably a parametrized test should be written to cover a reasonable set of inputs/expected outputs along with a code adjustment to allow the test to pass for the different cases.

I suppose I'll try to draft something shortly then.

**Comment by perimosocordiae** [comment]
Good catch @grlee77, I had expected the test cases in this PR to catch this inconsistency but I see now that they explicitly skip the `shape == (0, 0)` case for some reason.

**Pull request merged**